### PR TITLE
Search by name support in readGroupSets + refactor.

### PR DESCRIPTION
**Pull request by jeromekelleher**
- Refactored to make ID search fully hierarchical
- Added support for searching ReadGroupSets by name (issue #534)
- Generalised the simulator initialisation structures
- Removed multiple locations for initialising internal maps, preferring
  the use of addDataset, addVariantSet, etc.
- Fixed up some test cases that depended on brittle interfaces.

### CODE REVIEW

File: ga4gh/backend.py [hunk 1]
```diff
@@ -339,6 +354,13 @@ def _topLevelObjectGenerator(self, request, idMap, idList):
                 nextPageToken = str(currentIndex)
             yield object_.toProtocolElement(), nextPageToken
 
+    def _singleObjectGenerator(self, datamodelObject):
+        """
+        Returns a generator suitable for a search method in which the
+        result set is a single object.
+        """
+        return [(datamodelObject.toProtocolElement(), None)]
```
Comment (dcolligan) [inline ga4gh/backend.py]: this doesn't look like it returns a generator

Comment (jeromekelleher) [inline ga4gh/backend.py]: We can wrap an iterator around it if you like to force it to be a generator. Once it's an iterable it doesn't matter. Maybe I should change the comment to "returns an iterable..."?

Comment (dcolligan) [inline ga4gh/backend.py]: yeah, I wasn't concerned about semantics, more about clarity... if a comment says a method returns a generator and it doesn't, it's confusing

File: ga4gh/backend.py [hunk 2]
```diff
@@ -740,5 +765,4 @@ def __init__(self, dataDir):
             directory != referencesDirName]
         for datasetDir in datasetDirs:
             dataset = datasets.FileSystemDataset(datasetDir)
-            self._datasetIdMap[dataset.getId()] = dataset
-        self._datasetIds = sorted(self._datasetIdMap.keys())
```
Comment (dcolligan) [inline ga4gh/backend.py]: Are we throwing away the assumption that the datasets are in some kind of order here?  We control the dataset names in the simulated backend, but not in the file system backend.  Should we care?

File: ga4gh/datamodel/datasets.py [hunk 3]
```diff
@@ -130,9 +147,7 @@ def __init__(self, dataDir):
             if os.path.isdir(relativePath):
                 variantSet = variants.HtslibVariantSet(
                     self, localId, relativePath)
-                self._variantSetIdMap[variantSet.getId()] = variantSet
-        self._variantSetIds = sorted(self._variantSetIdMap.keys())
```
Comment (dcolligan) [inline ga4gh/datamodel/datasets.py]: same question about sorting here

File: ga4gh/datamodel/datasets.py [hunk 4]
```diff
@@ -140,8 +155,4 @@ def __init__(self, dataDir):
             if os.path.isdir(relativePath):
                 readGroupSet = reads.HtslibReadGroupSet(
                     self, localId, relativePath)
-                self._readGroupSetIdMap[readGroupSet.getId()] = readGroupSet
-                for readGroup in readGroupSet.getReadGroups():
-                    self._readGroupIdMap[readGroup.getId()] = readGroup
-        self._readGroupSetIds = sorted(self._readGroupSetIdMap.keys())
-        self._readGroupIds = sorted(self._readGroupIdMap.keys())
```
Comment (dcolligan) [inline ga4gh/datamodel/datasets.py]: and here

File: ga4gh/backend.py (hunk 2 above)
Comment (jeromekelleher) [inline ga4gh/backend.py]: I don't think it really matters. There's no requirement in the protocol that these objects be returned in any particular order, so we're not constrained by that. The only issue I can think of potentially causing problems is with paging over multiple processes. If (somehow) two different processes have these lists in different orders, and subsequent page requests get routed to different processes we'll have a bad time.

This seems quite unlikely, but perhaps it's possible. One option is to use the bisect module to insert the ids into the list in sorted order. Another would be to abstract out the file-system walking code and to make sure that files/directories are visited in sorted order. What do you think, is this worth worrying about?

**Comment by jeromekelleher** [comment]
@dcolligan, pushed an update. I'm not sure what to do about the sortedness requirement --- we can put it in now, or delay thinking about it properly to another day with an issue.

**Comment by dcolligan** [comment]
If what we should do isn't clear, than I have no problem +1'ing this.

Another consideration is if we want to have the ordering stable across server restarts.  I think that's a good property, but I also don't think it's a guarantee that the OS will return the directory names to us in the same order every time.  Something to think about when it becomes an issue...

**Pull request merged**